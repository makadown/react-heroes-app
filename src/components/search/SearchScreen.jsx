import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = '' } = queryString.parse(location.search);

  const [formValues, handleInputChange] = useForm({
    searchWord: q,
  });

  let { searchWord } = formValues;
  const filteredHeroes = useMemo(() => getHeroesByName(q), [q]);

  const handleSubmit = (e) => {
    e.preventDefault();
    // console.log(searchWord);
    history.push(`?q=${searchWord}`);
  };

  return (
    <div>
      <div className="row">
        <div className="col-5">
          <h4>Search</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <div className="form-group">
              <input
                type="text"
                name="searchWord"
                autoComplete="off"
                placeholder="Find your hero"
                onChange={handleInputChange}
                className="form-control"
                value={searchWord}
              />
            </div>
            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Search...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {q === '' && <div className="alert alert-info">Search a hero</div>}
          {q !== '' && filteredHeroes.length === 0 && (
            <div className="alert alert-info">No results found with {q}</div>
          )}

          {filteredHeroes.map((hero) => (
            <HeroCard key={hero.id} props={hero}></HeroCard>
          ))}
        </div>
      </div>
    </div>
  );
};
