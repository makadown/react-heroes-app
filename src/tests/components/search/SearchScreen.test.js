import React from 'react';
import { MemoryRouter, Route } from 'react-router';
import { mount } from 'enzyme';
import { SearchScreen } from '../../../components/search/SearchScreen';

describe('Pruebas con <SearchScreen/>', () => {

    test('debe mostrarse correctamente con valores por defecto ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen}>

                </Route>
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });

    test('debe mostrar a Batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen}>

                </Route>
            </MemoryRouter>
        );
        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    });
    
    test('debe mostrar error si no se encuentra el heroe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman_popo_']}>
                <Route path="/search" component={SearchScreen}>

                </Route>
            </MemoryRouter>
        ); 
        expect(wrapper.find('.alert-info').text().trim()).toContain('No results found');
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de llamar el push del history al enviar texto de formulario para buscar', () => {
        const historyMock = {
            push: jest.fn(),
            location: {},
            listen: jest.fn(),
            createHref: jest.fn(),
            replace: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path="/search" 
                    component={ () => <SearchScreen history={historyMock}/> }>

                </Route>
            </MemoryRouter>
        ); 
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault() {}
        });

        expect(historyMock.push).toHaveBeenCalledWith(`?q=batman`);
    });
    

});
