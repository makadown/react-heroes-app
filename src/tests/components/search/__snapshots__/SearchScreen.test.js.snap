// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con <SearchScreen/> debe mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchWord"
            onChange={[Function]}
            placeholder="Find your hero"
            type="text"
            value="batman"
          />
        </div>
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3 animate__animated animate__fadeIn text-white bg-dark"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                </small>
              </p>
              <a
                className="btn btn-info"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                More . . .
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas con <SearchScreen/> debe mostrar error si no se encuentra el heroe 1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchWord"
            onChange={[Function]}
            placeholder="Find your hero"
            type="text"
            value="batman_popo_"
          />
        </div>
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        No results found with 
        batman_popo_
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas con <SearchScreen/> debe mostrarse correctamente con valores por defecto  1`] = `
<div>
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            autoComplete="off"
            className="form-control"
            name="searchWord"
            onChange={[Function]}
            placeholder="Find your hero"
            type="text"
            value=""
          />
        </div>
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
